polling_stations/apps/data_importers/data_types.py

```diff
@@ -0,0 +1,291 @@
+        "area",
+        "polling_station_id",
+    ],
+)
+
+
+class CustomSet(metaclass=abc.ABCMeta):
+    def __init__(self):
+        self.elements = set()
+
+    def add(self, element):
+        self.elements.add(self.build_namedtuple(element))
+
+    @abc.abstractmethod
+    def build_namedtuple(self, element):
+        pass
+
+
+class AssignPollingStationsMixin(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def get_polling_station_lookup(self):
+        pass
+
+    @property
+    def council_id(self):  # TODO Deal with old_to_new council_ids map
+        for e in self.elements:
+            if isinstance(e, dict):
+                return e["council"].council_id
+            else:
+                return e.council.council_id
+
+    def update_uprn_to_council_model(self, polling_station_lookup=None):
+        if not polling_station_lookup:
+            polling_station_lookup = self.get_polling_station_lookup()
+
+        uprns_in_council = UprnToCouncil.objects.filter(lad=self.council_id)
+        seen = set()
+        for polling_station_id, uprns in polling_station_lookup.items():
+            uprns_in_council.filter(uprn__in=uprns).update(
+                polling_station_id=polling_station_id
+            )
+            if self.__class__.__name__ == "DistrictSet":
+                # We have to do this in case there are two districts which overlap
+                # and an address falls within that overlapping area.
+                duplicates = [u for u in uprns if u in seen]
+                uprns_in_council.filter(uprn__in=duplicates).update(
+                    polling_station_id=""
+                )
+                seen.update(uprns)
+
+
+class DistrictSet(CustomSet, AssignPollingStationsMixin):
+    def build_namedtuple(self, element):
+
+        # MultiPolygon is mutable, so we must serialize it to store in a tuple
+        area = element["area"].ewkb  # use ewkb so it encodes srid
+
+        return District(
+            element.get("name", ""),
+            element["council"],
+            element["internal_council_id"],
+            element.get("extra_id", ""),
+            area,
+            element.get("polling_station_id", ""),
+        )
+
+    def save(self):
+        districts_db = []
+        for district in self.elements:
+            record = PollingDistrict(
+                name=district.name,
+                council=district.council,
+                internal_council_id=district.internal_council_id,
+                extra_id=district.extra_id,
+                area=district.area,
+                polling_station_id=district.polling_station_id,
+            )
+            districts_db.append(record)
+        PollingDistrict.objects.bulk_create(districts_db)
+
+    def get_polling_station_lookup(self):
+        """
+        for each address, build a lookup of polling_station_id -> set of uprns
+        This does the lookup in the db so you need to have called save() before
+        using this.
```

Comment:
We could make this less error prone by having a 'saved' flag, so we do something like (cut down example for brevity):

```py
class CustomSet(metaclass=abc.ABCMeta):
    def __init__(self):
        self.elements = set()
        self.saved = False

    def add(self, element):
        self.elements.add(self.build_namedtuple(element))
        self.saved = False

class DistrictSet(CustomSet):

    def save(self):
        # do stuff
        self.saved = True

    def get_polling_station_lookup(self):
        if not self.saved:
            raise Exception('oh noes!!')
        # do stuff
```

then we can't mess up by calling `get_polling_station_lookup()` before `save()`.

Reply:
Nice, I'll build that in.